js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,385 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2017 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ws
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dop251/goja"
+	"github.com/gorilla/websocket"
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/lib/metrics"
+	"github.com/loadimpact/k6/stats"
+)
+
+type WS struct{}
+
+type Socket struct {
+	ctx           context.Context
+	conn          *websocket.Conn
+	eventHandlers map[string][]goja.Callable
+	scheduled     chan goja.Callable
+	done          chan struct{}
+	shutdownOnce  sync.Once
+
+	msgsSent, msgsReceived float64
+
+	pingSendTimestamps map[string]time.Time
+	pingSendCounter    int
+	pingDeltas         []time.Duration
+}
+
+const writeWait = 10 * time.Second
+
+var (
+	newline = []byte{'\n'}
+	space   = []byte{' '}
+)
+
+func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) *http.Response {
```

Comment:
Oh, I just noticed this returns a raw HTTP response, which I wouldn't recommend. If anything, wrap it in a response struct from the `k6/http` module.

Also: you can return `error` or `(type, error)` from **top-level** functions in a module and have them automatically turn into throws. (Won't work with objects you create, that kind of bridge is sadly too slow to create in a hot path...)